Updated README with API spec and description of application (#2)
cul/ldpd-uri-service: README.md

# URI Service (Application)

**URI Service** is a standalone Rails 5 application that creates/stores local and temporary terms and caches external URI terms. To start the application's only interface will be a JSON api, though we see expanding the application to include html pages for local terms.

URI Service makes some assumptions about how URIs should be organized. A URI is always related to a vocabulary. A vocabulary has many URI terms. Multiple authorities might be represented within a vocabulary. A URI must be unique to a vocabulary, but a URI can appear in multiple vocabularies.

> In order to use the JSON api, URI Service requires token authentication. URI Service does not provide granular authorization. The application that uses the api should control authorization in anyway it deems necessary.

Reviewer: ```suggestion
In order to use the JSON api, URI Service requires token authentication. URI Service does not provide granular authorization. The application that uses the api should control authorization in any way it deems necessary.
```

Used when you want to add a value to your URI Service datastore, but have no authority information about the term or do not wish to create a local URI. No two temporary terms within the same vocabulary can have the same value. Basically, a temporary term is intended to identify an exact string value rather than identifying an intellectual entity. Temporary terms should eventually be replaced by external or local terms later on when more information is known about the entity to which you are referring . Example: We want to record information about the author of an old and mysterious letter by "John Smith." We don't know which "John Smith," this refers to, so we'll create (or re-use) a temporary URI that's associated with the value "John Smith."

#### Custom Fields
TODO

## Installation
TODO

## Services (job queues, cache servers, search engines, etc.)
URI Services stores local, temporary and external terms in a mysql database. Solr is used to index terms for fast lookup and searching. Redis/Resque will be used to run scheduled jobs.

## Deployment instructions
TODO

## API v1

### Vocabularies
#### List Vocabularies
- **Request**
  `GET /vocabularies`

- **Success Response**

>   Sorts vocabularies alphabetically by label. Defaults to `page = 1` and `per_page = 20`.

Reviewer: We should update the vocabularies response example below to have the terms sorted alphabetically by label.

  _Status:_ 200

  _Body:_

  ```json
  {
    "page": 1,
    "per_page": 10,
    "total_results": 2,
    "vocabularies": [
      { "string_key": "name" , "label": "Names" },
      { "string_key": "collections" , "label": "Collections" }
    ]
  }
  ```

#### Find Single Vocabulary
- **Request**

  `GET /vocabularies/:string_key`

- **Successful Response**

  _Status:_ 200

  _Body:_
  ```json
  {
     "string_key": "names",
     "label": "Names"
  }
  ```

#### Create Vocabulary
- **Request**

  `POST /vocabularies`

>   `string_key` is a required parameter.

Reviewer: `label` is also a required parameter for a new vocabulary.

Author: should this also be required in the model? we had not talked about this requirement

Reviewer: you're right, it should

- **Success Response**

  _Status:_ `201`

  _Body:_
    ```json
    {
      "string_key": "names",
      "label": "Names"
    }
    ```

> #### Update Vocabulary

Reviewer: Would be good to mention here that the only property of a vocabulary that you're allowed to update is the `label` (and `string_key` should never change).

- **Request**

  `PATCH /vocabularies/:string_key`

- **Success Response**

  _Status:_ 200

  _Body:_
  ```json
  {
     "string_key": "names",
     "label": "Names"
  }
  ```

#### Delete Vocabulary
- **Request**

  `DELETE /vocabularies/:string_key`

- **Success Response**

  _Status:_ 204

  _Body:_ nil

- **Error Response**

>   - If resource does not exists returns `404`

Reviewer: ```suggestion
  - If resource does not exist returns `404`
```

  - If not successful return `500`

#### Send back description/instruction
- **Request**

  `OPTIONS /vocabularies/`
- **Success Response**
  TODO

> ### Vocabularies/Term

Reviewer: ```suggestion
### Vocabularies/Terms
```

#### List/Search through Terms in Vocabulary
- **Request**

  `GET /vocabularies/:string_key/terms`

  Examples:
  ```
  GET /vocabularies/:string_key/terms?page=1&per_page=20`
  GET /vocabularies/:string_key/terms?q=smith&authority=naf&page=1&per_page=20`
  GET /vocabularies/:string_key/terms?uri=https://creativecommons.org/publicdomain/zero/1.0/

>   GET /vocabularies/:string_key/terms?value=Smith # returns temp term b/c no uri was given

Reviewer: ```suggestion
  GET /vocabularies/:string_key/terms?label=Smith # returns temp term b/c no uri was given
```

  ```

  Allowed query fields:

>     - `q`: query uri terms

Reviewer: ```suggestion
    - `q`: query term labels using fuzzy matching
```

    - `authority`: facet by authority

>     - `uri`: search for uri

Reviewer: ```suggestion
    - `uri`: search by exact uri string
```

>     - `value`: search for value with in temp terms

Reviewer: ```suggestion
    - `label`: search by exact label string
```

> 

Reviewer: ```suggestion
- `starts_with`: return terms that start with the given string

```
Should this be a case insensitive search?

  Other allowed params:
    - `per_page`: number of terms per page, default: 20
    - `page`: page to display

- **Success Response**

>   Sorting alphabetically by ????

Reviewer: ```suggestion
  Sorted primarily by match score (in the case of a q search), secondarily by label alphabetical sort
```

Reviewer: FYI: by "match score" I mean by whatever order solr gives us when it returns results. on the solr side, we'll primarily sort by score, and the second sort comparison will be alphabetical label sort.

Author: gotcha

  _Status:_ 200

  _Body:_
  ```json
  {
    "page": 1,
    "per_page": 25,
    "total records": 55,
    "terms": [
      {
        "pref_label": 'Carla Galarza',
        "uri":"https://id.library.columbia.edu/term/1111-2222-3333-...",
        "authority": "local",
        "custom_field_1": "something"
      },
      {

>         "pref_label": Eric O'Hanlon',

Reviewer: ```suggestion
        "pref_label": 'Eric O'Hanlon',
```

        "uri":"https://id.library.columbia.edu/term/4444-5555-6666...",
        "authority": "local",
        "custom_field_2": "another thing"
      }
    ]
  }
  ```

#### Create term
- **Request**

  `POST /vocabularies/:string_key/terms`

>   _Required params:_ term_type, label, uri

Reviewer: ```suggestion
  _Required params:_ term_type, label, uri (unless term_type = local or temporary)
```

>   _Optional params:_ Any vocabulary-specific custom fields

Reviewer: ```suggestion
  _Optional params:_ authority, and any vocabulary-specific custom fields
```

- **Success Response**

  _Status:_ 201

  _Body:_ json representation of object

#### Update Term
- **Request**

  `PATCH /vocabularies/:string_key/terms/:uri`

>   URI should be encoded

Reviewer: ```suggestion
  URI should be URL-encoded
```

  Can only change label and custom fields. URI and term_type can only be added with a create.

- **Success Response**

  _Status:_ 200

  _Body:_ JSON representation of object

- **Error Response**
  - return 404 (if object cannot be found)

#### Delete term
- **Request**

  `DELETE  /vocabularies/:string_key/terms/:uri`

>   URI should be encoded.

Reviewer: ```suggestion
  URI should be URL-encoded.
```

- **Success Response**

  _Status:_ 200

  _Body:_ none

### Vocabularies/Terms

#### List/Search through Terms in Vocabulary
- **Request**

  `GET /vocabularies/:string_key/terms`

  Examples:
  ```
  GET /vocabularies/:string_key/terms?page=1&per_page=20`
  GET /vocabularies/:string_key/terms?q=smith&authority=naf&page=1&per_page=20`
  GET /vocabularies/:string_key/terms?uri=https://creativecommons.org/publicdomain/zero/1.0/
  GET /vocabularies/:string_key/terms?label=Smith # returns temp term b/c no uri was given
  ```

  Allowed query fields:
    - `q`: query term labels using fuzzy matching
    - `authority`: facet by authority
    - `uri`: search by exact uri string
    - `label`: search by exact label string

  Other allowed params:
    - `per_page`: number of terms per page, default: 20
    - `page`: page to display

- **Success Response**

  Sorted primarily by match score (in the case of a q search), secondarily by label alphabetical sort

  _Status:_ 200

  _Body:_
  ```json
  {
    "page": 1,
    "per_page": 25,
    "total records": 55,
    "terms": [
      {

>         "pref_label": 'Carla Galarza',

Reviewer: ```suggestion
        "pref_label": "Carla Galarza",
```

Reviewer: Github was highlighting this value in red, so maybe it prefers double quotes.

        "uri":"https://id.library.columbia.edu/term/1111-2222-3333-...",
        "authority": "local",
        "custom_field_1": "something"
      },
      {

>         "pref_label": 'Eric O'Hanlon',

Reviewer: ```suggestion
        "pref_label": "Eric O'Hanlon",
```

Reviewer: Github was also highlighting this value in red, so maybe it prefers double quotes.  Plus, double quotes around the name string are better for the apostrophe in my name, haha.
